client/common/sso_mib_tokens.c

```diff
@@ -0,0 +1,199 @@
+	if (!kid)
+	{
+		goto cleanup;
+	}
+
+	params = mib_pop_params_new(MIB_AUTH_SCHEME_POP, MIB_REQUEST_METHOD_GET, "");
+	mib_pop_params_set_kid(params, kid);
+	MIBPrt* prt = mib_client_app_acquire_token_interactive(
+	    client_context->mibClientWrapper->app, scopes, MIB_PROMPT_NONE, NULL, NULL, NULL, params);
+	if (prt)
+	{
+		*token = strdup(mib_prt_get_access_token(prt));
+		rc = TRUE;
+		g_object_unref(prt);
+	}
+
+cleanup:
+	if (params)
+		g_object_unref(params);
+	WINPR_JSON_Delete(json);
+	free(req_cnf_dec);
+	g_slist_free_full(scopes, g_free);
+	return rc;
+}
+
+BOOL sso_mib_get_access_token(rdpContext* context, AccessTokenType tokenType, char** token,
+                              size_t count, ...)
+{
+	BOOL rc = FALSE;
+	rdpClientContext* client_context = (rdpClientContext*)context;
+	WINPR_ASSERT(client_context);
+	WINPR_ASSERT(client_context->mibClientWrapper);
+
+	if (!client_context->mibClientWrapper->app)
+	{
+		const char* client_id =
+		    freerdp_settings_get_string(context->settings, FreeRDP_GatewayAvdClientID);
+		client_context->mibClientWrapper->app =
+		    mib_public_client_app_new(client_id, MIB_AUTHORITY_COMMON, NULL, NULL);
+	}
+
+	if (!client_context->mibClientWrapper->app)
+		;
+	return ERROR_INTERNAL_ERROR;
+
+	const char* scope = NULL;
+	const char* req_cnf = NULL;
+
+	va_list ap;
+	va_start(ap, count);
+
+	if ((client_context->mibClientWrapper->state == SSO_MIB_STATE_INIT) ||
+	    (client_context->mibClientWrapper->state == SSO_MIB_STATE_SUCCESS))
+	{
+		switch (tokenType)
+		{
+			case ACCESS_TOKEN_TYPE_AVD:
+			{
+				rc = sso_mib_get_avd_access_token(client_context, token);
+				if (rc)
+					client_context->mibClientWrapper->state = SSO_MIB_STATE_SUCCESS;
+				else
+				{
+					WLog_WARN(TAG, "Getting AVD token from identity broker failed, falling back to "
+					               "browser-based authentication.");
+					client_context->mibClientWrapper->state = SSO_MIB_STATE_FAILED;
+				}
+			}
+			break;
+			case ACCESS_TOKEN_TYPE_AAD:
+			{
+				scope = va_arg(ap, const char*);
+				req_cnf = va_arg(ap, const char*);
+				// Setup scope without URL encoding for sso-mib
+				char* scope_copy = winpr_str_url_decode(scope, strlen(scope));
+				if (!scope_copy)
+					WLog_ERR(TAG, "Failed to decode scope");
+				else
+				{
+					rc =
+					    sso_mib_get_rdsaad_access_token(client_context, scope_copy, req_cnf, token);
+					free(scope_copy);
+					if (rc)
+						client_context->mibClientWrapper->state = SSO_MIB_STATE_SUCCESS;
+					else
+					{
+						WLog_WARN(TAG,
+						          "Getting RDS token from identity broker failed, falling back to "
+						          "browser-based authentication.");
+						client_context->mibClientWrapper->state = SSO_MIB_STATE_FAILED;
+					}
+				}
+			}
+			break;
+			default:
+				break;
+		}
+	}
+	if (!rc && client_context->mibClientWrapper->GetCommonAccessToken)
+		rc = client_context->mibClientWrapper->GetCommonAccessToken(context, tokenType, token,
```

Comment:
Does this needs to call context->instance->GetAccessToken(context->instance, [...])`?

Additionally, we need to move error checking around here as otherwise `scope` and `req_cnf` would be NULL in case the first request fails (and sets the `state` to `SSO_MIB_STATE_FAILED`).

Reply:
it does call this.
* The `GetCommonAccessToken` saved in the context is the one set before, so `freerdp_common_context`
* `scope` and `req_cnf` are not `NULL`, they are extracted from the `va_list` if `` is `ACCESS_TOKEN_TYPE_AAD`

Reply:
ah, ok, now I see, did miss that case :grin: